### ISSUE TITLE
ENH: implicit args in blackscholes

### ISSUE THREAD

**Issue by masanorihirano** [pfnet-research/pfhedge#485]
Currently, BlackScholes modules work like:
```
from pfhedge.instruments import BrownianStock
from pfhedge.instruments import EuropeanOption
from pfhedge.nn import BlackScholes

derivative = EuropeanOption(BrownianStock())
pricer = lambda derivative: BlackScholes(derivative).price(
             log_moneyness=derivative.log_moneyness(),
             time_to_maturity=derivative.time_to_maturity(),
             volatility=derivative.ul().volatility)
derivative = EuropeanOption(BrownianStock(), maturity=5/250)
derivative.list(pricer, cost=1e-4)
derivative.simulate(n_paths=2)
derivative.spot
```

However, the part,
```
pricer = lambda derivative: BlackScholes(derivative).price(
             log_moneyness=derivative.log_moneyness(),
             time_to_maturity=derivative.time_to_maturity(),
             volatility=derivative.ul().volatility)
```
seems a little bit redundant and not universal.
I think it comes from the following reasons:
 1. it is obvious that each args are generated from derivatives.
 2. required args (and the number of required args) depend on underliers

Thus, I suggest that all args are not required, and if the args are missing, these required args are calculated automatically based on registered derivatives.

Of course, in the case of derivative using autogreeks, these implicit args cannot be applied, but, I think my suggestion is worth being considere.

This suggestion is not limited to BlackScholes.price, but also to all other methods in greeks methods.

**Comment by masanorihirano** [comment]
Note that the current best workaround for the universality problem is
```
pricer = lambda derivative: BlackScholes(derivative).price(
             *[get_feature(f).of(derivative=derivative) for f in BlackScholes(derivative).inputs()])
```

**Comment by simaki** [comment]
Sounds great 👍 
Maybe I will make that enhancement when I find the time, while it is more than welcome if you kindly send a patch.
If you want to implement this let me know to prevent duplicate work.

One way is to register default arguments as attributes of a module in __init__ and pass them in function calls as:

```py
    def __init__(self, ..., log_moneyness=None):
        self.log_moneyness = log_moneyness

    def price(self, log_moneyness=None, ...):
        log_moneyness = self.log_moneyness or log_moneyness
        ...
```

**Comment by masanorihirano** [comment]
I know it is not high priority problem for you and me. So, I wrote this as just a note.
I'll let you know here before I do this implementation.

**Comment by masanorihirano** [comment]
I'll make a prototype

**Comment by masanorihirano** [comment]
I made a prototype for this: #516 
Currently, this is only implemented for European option.
If it looks good for you (@simaki ), I'll implement to the others.

**Comment by simaki** [comment]
Resolved by #516
